### PR TITLE
Always evaluate the finder in `driver.waitFor*()`

### LINKED ISSUE #11327: FlutterDriver.waitForAbsent() never completing
#### Problem

Tried to use this inside Google today, and getting incorrect behavior:

```dart
await driver.waitForAbsent(find.text('foo'), timeout: const Duration(seconds: 20));
```

* If I call that, the future never completes, and the test fails with a timeout exception after 20 seconds.
* If I *first* await a delayed future of 5 seconds to let the Snackbar that holds the text go away, *then* run the `await` above, then the future completes just fine, and the test succeeds.

#### Analysis

The future is never completing because of an error during this call in `extension.dart`:
```dart
await _waitUntilFrame(() => !finder.precache());
```

The error is:
```
'package:flutter_test/src/finders.dart': Failed assertion: line 269: '_cachedResult == null': is not true.
```

It seems weird that it's causing the future to never complete rather than completing with an error...

### PR DESCRIPTION
**Pull request by yjbanov**
I added tests on top of @tvolkert's changes from https://github.com/flutter/flutter/pull/11329.

Fixes #11327

cc @Hixie for review

### CODE REVIEW

File: dev/integration_tests/ui/test_driver/driver_test.dart [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    test('waitFor should find text "present"', () async {
+      await driver.waitFor(presentText);
+    });
+
+    test('waitForAbsent should time out waiting for text "present" to disappear', () async {
+      try {
+        await driver.waitForAbsent(presentText, timeout: const Duration(seconds: 1));
+        fail('expected DriverError');
+      } on DriverError catch(error) {
```
Comment (Hixie) [inline dev/integration_tests/ui/test_driver/driver_test.dart]: nittiest of nits: space after `catch` before `(`

**Comment by Hixie** [comment]
RSLGTM

File: dev/integration_tests/ui/test_driver/driver_test.dart (hunk 1 above)
Comment (yjbanov) [inline dev/integration_tests/ui/test_driver/driver_test.dart]: Done.

**Comment by Hixie** [comment]
For those looking at performance benchmarks:

This patch was the first patch immediately after we changed the test hardware from Nexus 5s to Moto G4s. We later tested removing this patch to see if it affected the numbers and it did not.

**Pull request merged**